src/lowtime_graph.rs

```diff
@@ -0,0 +1,214 @@
+use std::collections::{HashMap, HashSet};
+use ordered_float::OrderedFloat;
+use pathfinding::directed::edmonds_karp::{
+    SparseCapacity,
+    Edge,
+    EKFlows,
+    edmonds_karp,
+};
+
+
+#[derive(Clone)]
+pub struct LowtimeGraph {
+    node_ids: Vec<u32>,
+    source_node_id: Option<u32>,
+    sink_node_id: Option<u32>,
+    edges: HashMap<u32, HashMap<u32, LowtimeEdge>>,
+    preds: HashMap<u32, HashSet<u32>>,
+    num_edges: usize,
+}
+
+impl LowtimeGraph {
+    pub fn new() -> Self {
+        LowtimeGraph {
+            node_ids: Vec::new(),
+            source_node_id: None,
+            sink_node_id: None,
+            edges: HashMap::new(),
+            preds: HashMap::new(),
+            num_edges: 0,
+        }
+    }
+
+    pub fn of_python(
+        mut node_ids: Vec<u32>,
+        source_node_id: u32,
+        sink_node_id: u32,
+        edges_raw: Vec<((u32, u32), (f64, f64, f64, f64))>,
+    ) -> Self {
+        let mut graph = LowtimeGraph::new();
+        node_ids.sort();
+        graph.node_ids = node_ids.clone();
+        graph.source_node_id = Some(source_node_id);
+        graph.sink_node_id = Some(sink_node_id);
+
+        edges_raw.iter().for_each(|(
+                (from, to),
+                (capacity, flow, ub, lb),
+            )| {
+            graph.add_edge(*from, *to, LowtimeEdge::new(
+                OrderedFloat(*capacity),
+                OrderedFloat(*flow),
+                OrderedFloat(*ub),
+                OrderedFloat(*lb),
+            ))
+        });
+        graph
+    }
+
+    pub fn max_flow(&self) -> EKFlows<u32, OrderedFloat<f64>> {
+        let edges_edmonds_karp = self.get_ek_preprocessed_edges();
+        let (flows, max_flow, min_cut) = edmonds_karp::<_, _, _, SparseCapacity<_>>(
+            &self.node_ids,
+            &self.source_node_id.unwrap(),
+            &self.sink_node_id.unwrap(),
+            edges_edmonds_karp,
+        );
+        (flows, max_flow, min_cut)
+    }
+
+    pub fn get_source_node_id(&self) -> u32 {
+        self.source_node_id.unwrap()
+    }
```

Comment:
I went ahead and made all the "trivial" vars public in `LowtimeGraph` and `LowtimeEdge`. The only things not made `pub` are:
* `LowtimeGraph.edges`
* `LowtimeGraph.preds`
* `LowtimeGraph.num_nodes`
* `LowtimeGraph.num_edges`
For `edges` and `preds`, I think that explicitly managing mutability will somehow make it easier to write future code.
For `num_nodes` and `num_edges`, `num_nodes` is not stored as a member var (computed on the fly) while `num_edges` is (hard-coded count); I think it would be weird to have `num_nodes` as a function by `num_edges` as a member var access. I did this in the first place because I thought the number of edges would be significantly larger than number of nodes, so computing num edges on the fly would take much longer than computing num nodes.

Reply:
All makes sense! `edges` and `preds` ought to be private.